light-rpc/src/bridge.rs

```diff
@@ -0,0 +1,181 @@
+
+        let signature = bincode::deserialize::<Transaction>(&wire_transaction)?.signatures[0];
+        let signature = BinaryEncoding::Base58.encode(signature);
+
+        let conn = self.connection_cache.get_connection(&self.tpu_addr);
+        conn.send_wire_transaction_async(wire_transaction.clone())?;
+
+        match max_retries.unwrap_or(5) {
+            0 => (),
+            max_retries => {
+                tokio::spawn(async move {
+                    let mut interval = tokio::time::interval(Duration::from_millis(10));
+
+                    for _ in 0..max_retries {
+                        interval.tick().await;
+
+                        if let Err(_err) =
+                            conn.send_wire_transaction_async(wire_transaction.clone())
+                        {
+                            break;
+                        }
+                    }
+                });
+            }
+        }
+
+        Ok(signature)
+    }
+
+    /// Serialize params and execute the specified method
+    pub async fn execute_rpc_request(
+        &self,
+        method: RpcMethod,
+    ) -> Result<serde_json::Value, JsonRpcError> {
+        match method {
+            RpcMethod::SendTransaction(transaction, config) => {
+                Ok(self.send_transaction(transaction, config)?.into())
+            }
+        }
+    }
+
+    /// List for `JsonRpc` requests
+    pub async fn start_server(self, addr: impl ToSocketAddrs) -> Result<(), std::io::Error> {
+        let bridge = Arc::new(self);
+
+        let json_cfg = web::JsonConfig::default().error_handler(|err, req| {
+            let err = JsonRpcRes::Err(serde_json::Value::String(format!("{err}")))
+                .respond_to(req)
+                .into_body();
+            actix_web::error::ErrorBadRequest(err)
+        });
+
+        HttpServer::new(move || {
+            App::new()
+                .app_data(web::Data::new(bridge.clone()))
+                .app_data(json_cfg.clone())
+                .route("/", web::post().to(Self::rpc_route))
+        })
+        .bind(addr)?
+        .run()
+        .await
+    }
+
+    async fn rpc_route(
+        json_rpc_req: web::Json<RpcMethod>,
+        state: web::Data<Arc<LightBridge>>,
+    ) -> JsonRpcRes {
+        state
+            .execute_rpc_request(json_rpc_req.0)
+            .await
+            .try_into()
+            .unwrap()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use solana_sdk::{
+        message::Message, native_token::LAMPORTS_PER_SOL, pubkey::Pubkey, signature::Keypair,
+        signer::Signer, system_instruction, transaction::Transaction,
+    };
+
+    use crate::{bridge::LightBridge, encoding::BinaryEncoding};
+
+    const RPC_ADDR: &str = "127.0.0.1:8899";
+    const TPU_ADDR: &str = "127.0.0.1:1027";
+    const CONNECTION_POOL_SIZE: usize = 1;
+
+    #[test]
+    fn test_send_transaction() {
+        let light_bridge = LightBridge::new(
```

Comment:
These tests do not really test the http server. You should write a tests which really tests the http server instead of calling thin_client in light_bridge.

Reply:
We can't really do that now, until and unless we implement the methods required to make RpcClient work.
I'll create a new issue for this

Reply:
Followup issue https://github.com/blockworks-foundation/solana/issues/13